packages/grafana-llm-app/pkg/mcp/live_server.go

```diff
@@ -0,0 +1,218 @@
+func (s *GrafanaLiveServer) HandleMessage(ctx context.Context, req *backend.PublishStreamRequest) error {
+	// Get the session from the sessions map.
+	sessionI, ok := s.sessions.Load(req.Path)
+	if !ok {
+		return ErrStreamNotFound
+	}
+	session := sessionI.(*liveSession)
+
+	// Modify the context if a context function is set.
+	if s.contextFunc != nil {
+		ctx = s.contextFunc(ctx, &req.PluginContext)
+	}
+
+	// Process the message through the MCPServer.
+	response := s.server.HandleMessage(ctx, req.Data)
+
+	// Only send response if there is one (not for notifications).
+	if response != nil {
+		// Marshal the response to JSON. Errors should be impossible since we've
+		// just unmarshalled from a JSON-RPC message.
+		eventData, _ := json.Marshal(response)
+		return session.sender.SendJSON(eventData)
+	} else {
+		// For notifications, just send nil.
+		return session.sender.SendBytes(nil)
+	}
+}
+
+// ExtractClientFromGrafanaLiveRequest is a GrafanaLiveContextFunc which extracts the Grafana config
+// from settings and sets the client in the context.
+func extractGrafanaClientFromGrafanaLiveRequest(ctx context.Context, pCtx *backend.PluginContext) context.Context {
+	t := client.DefaultTransportConfig()
+
+	cfg := backend.GrafanaConfigFromContext(ctx)
+	if cfg == nil {
+		return ctx
+	}
+	urlS, err := cfg.AppURL()
+	if err != nil {
+		return ctx
+	}
+	url, err := url.Parse(urlS)
+	if err != nil {
+		return ctx
+	}
+	if url.Host != "" {
+		t.Host = url.Host
+	}
+	// The Grafana client will always prefer HTTPS even if the URL is HTTP,
+	// so we need to limit the schemes to HTTP if the URL is HTTP.
+	if url.Scheme == "http" {
+		t.Schemes = []string{"http"}
+	}
+
+	// TODO: fetch ID token / auth token from headers, as the app client secret
+	// uses the plugin's service account, not the current user.
+	// Tracked in https://github.com/grafana/grafana-llm-app/issues/593.
+	if apiKey, err := cfg.PluginAppClientSecret(); err == nil {
+		t.APIKey = apiKey
+	}
+
+	c := client.NewHTTPClientWithConfig(strfmt.Default, t)
+	return mcpgrafana.WithGrafanaClient(ctx, c)
+}
+
+func extractGrafanaInfoFromGrafanaLiveRequest(ctx context.Context, pCtx *backend.PluginContext) context.Context {
+	cfg := backend.GrafanaConfigFromContext(ctx)
+	if cfg == nil {
+		return ctx
+	}
+	url, err := cfg.AppURL()
+	if err != nil {
+		return ctx
+	}
+	apiKey, _ := cfg.PluginAppClientSecret()
+	return mcpgrafana.WithGrafanaAPIKey(mcpgrafana.WithGrafanaURL(ctx, url), apiKey)
+}
+
+func extractIncidentClientFromGrafanaLiveRequest(ctx context.Context, pCtx *backend.PluginContext) context.Context {
+	cfg := backend.GrafanaConfigFromContext(ctx)
+	if cfg == nil {
+		return ctx
+	}
+	grafanaURL, err := cfg.AppURL()
+	if err != nil {
+		return ctx
+	}
+	apiKey, _ := cfg.PluginAppClientSecret()
+	incidentUrl := fmt.Sprintf("%s/api/plugins/grafana-incident-app/resources/api/", strings.TrimSuffix(grafanaURL, "/"))
```

Comment:
We may need to update this to `grafana-irm-app` at some point but it's OK for now.